src/Npgsql/Internal/PgSerializerOptions.cs

```diff
@@ -0,0 +1,147 @@
+using System;
+using System.Runtime.CompilerServices;
+using System.Text;
+using Npgsql.Internal.Postgres;
+using Npgsql.NameTranslation;
+using Npgsql.PostgresTypes;
+
+namespace Npgsql.Internal;
+
+public class PgSerializerOptions
+{
+    /// <summary>
+    /// Used by GetSchema to be able to attempt to resolve all type catalog types without exceptions.
+    /// </summary>
+    [field: ThreadStatic]
+    internal static bool IntrospectionCaller { get; set; }
+
+    readonly Func<string>? _timeZoneProvider;
+    readonly object _typeInfoCache;
+
+    internal PgSerializerOptions(NpgsqlDatabaseInfo typeCatalog, Func<string>? timeZoneProvider = null)
+    {
+        _timeZoneProvider = timeZoneProvider;
+        TypeCatalog = typeCatalog;
+        UnknownPgType = typeCatalog.GetPostgresTypeByName("unknown");
+        _typeInfoCache = PortableTypeIds ? new TypeInfoCache<DataTypeName>(this) : new TypeInfoCache<Oid>(this);
+    }
+
+    // Represents postgres unknown type, which can be used for reading and writing arbitrary text values.
+    public PostgresType UnknownPgType { get; set; }
+
+    // Used purely for type mapping, where we don't have a full set of types but resolvers might know enough.
+    readonly bool _introspectionInstance;
+    internal bool IntrospectionMode
+    {
+        get => _introspectionInstance || IntrospectionCaller;
+        init => _introspectionInstance = value;
+    }
+
+    /// Whether options should return a portable identifier (data type name) to prevent any generated id (oid) confusion across backends, this comes with a perf penalty.
+    internal bool PortableTypeIds { get; init; }
+    internal NpgsqlDatabaseInfo TypeCatalog { get; }
```

Comment:
At least in theory, NpgsqlDatabaseInfo is supposed to provide more information than just about types (e.g. supported features or whatever).Even if this property is currently used only as a type catalog, should it be named NpgsqlDatabaseInfo (or something else)?

Reply:
Yes potentially, though regardless I'm not fully happy with how database conditional mappings ended up looking (take a look at the range resolver)